midi_rag_system.py

```
@@ -0,0 +1,616 @@
+            'pattern_clusters': self.pattern_clusters
+        }
+
+        with open(file_path, 'wb') as f:
+            pickle.dump(data, f)
+        print(f"Knowledge base saved to {file_path}")
+
+    def load_knowledge_base(self, file_path: str = "midi_knowledge_base.pkl") -> bool:
+        """지식 베이스 로드"""
+        try:
+            with open(file_path, 'rb') as f:
+                data = pickle.load(f)
```

Ruff (0.12.2)

`pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)